types/src/generator/type.ts

```diff
@@ -0,0 +1,328 @@
+import assert from "assert";
+import {
+  ArrayType,
+  BuiltinType_Type,
+  JsgImplType_Type,
+  MaybeType,
+  NumberType,
+  Structure,
+  StructureType,
+  Type,
+  Type_Which,
+} from "@workerd/jsg/rtti.capnp.js";
+import ts, { factory as f } from "typescript";
+import { printNode } from "../print";
+
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findLastIndex
+export function findLastIndex<T>(
+  array: T[],
+  predicate: (value: T, index: number, array: T[]) => unknown
+) {
+  for (let i = array.length - 1; i >= 0; i--) {
+    if (predicate(array[i], i, array)) return i;
+  }
+  return -1;
+}
+
+// If `typeNode` has the shape `T | undefined`, returns `T`, otherwise returns
+// `undefined`.
+export function maybeUnwrapOptional(
```

Comment:
Do we really need to consult generated types to glimpes information from the original structure?